File: data-structures/queue/queue.hs
module Queue where

type Queue a = [a]

enqueue :: Queue a -> a -> Queue a
enqueue [] y = [y] 
enqueue xs y = xs ++ [y]

enqueueBatch :: Queue a -> [a] -> Queue a
enqueueBatch [] ys = ys
enqueueBatch xs ys = xs ++ ys

dequeue :: Queue a -> Queue a
dequeue []     = error "Queue is empty"
dequeue (x:xs) = xs

dequeueBatch :: Queue a -> Int -> Queue a
dequeueBatch xs     0 = xs
dequeueBatch (x:xs) n = if n <= length(x:xs) 
                           then dequeueBatch xs (n - 1)
                           else error "Cannot dequeue more items that the queue contains"

peek :: Queue a -> a
peek []     = error "Queue is empty"
peek (x:xs) = x

File: data-structures/queue/queue.py
class Queue:
    
    def __init__(self):
        self.data = []
    
    def peek(self):
        return self.data[0]

    def enqueue(self, x):
        self.data = self.data + [x]

    def enqueueBatch(self, x):
        if not isinstance(x, list):
            raise TypeError("Can only add lists with enqueueBatch")
        self.data = self.data + x

    def dequeue(self):
        self.data.pop(0)

    def dequeueBatch(self, n):
        if not isinstance(n, int):
            raise TypeError("Must remove an integer value from the queue")
        if n > len(self.data):
            raise IndexError("Cannot remove en element from an empty queue")
        while n>0:
            self.data.pop(0)
            n -= 1

